master/internal/rm/multirm/multirm.go

```diff
@@ -0,0 +1,358 @@
+package multirm
+
+import (
+	"crypto/tls"
+	"fmt"
+
+	"github.com/labstack/echo/v4"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/maps"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/determined-ai/determined/master/internal/config"
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/internal/rm"
+	"github.com/determined-ai/determined/master/internal/rm/agentrm"
+	"github.com/determined-ai/determined/master/internal/rm/kubernetesrm"
+	"github.com/determined-ai/determined/master/internal/rm/rmerrors"
+	"github.com/determined-ai/determined/master/internal/sproto"
+	"github.com/determined-ai/determined/master/pkg/aproto"
+	"github.com/determined-ai/determined/master/pkg/command"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/proto/pkg/apiv1"
+	"github.com/determined-ai/determined/proto/pkg/jobv1"
+)
+
+// ErrRMNotDefined returns a detailed error if a resource manager isn't found in the MultiRMRouter map.
+func ErrRMNotDefined(rm string) error {
+	return fmt.Errorf("resource manager %s not defined", rm)
+}
+
+// MultiRMRouter tracks all resource managers in the system.
+type MultiRMRouter struct {
+	defaultRMName string
+	rms           map[string]rm.ResourceManager
+	syslog        *logrus.Entry
+}
+
+// New creates a new resource manager with the master set-up.
+// Since multirm imports the rm module, we have to put the New() func
+// here to avoid an import cycle. If only one rmConfig is passed, return a
+// single Agent or Kubernetes RM.
+func New(
+	db *db.PgDB,
+	echo *echo.Echo,
+	rmConfigs []*config.ResourceManagerWithPoolsConfig,
+	tcd *model.TaskContainerDefaultsConfig,
+	opts *aproto.MasterSetAgentOptions,
+	cert *tls.Certificate,
+) rm.ResourceManager {
+	if len(rmConfigs) <= 1 {
+		config := rmConfigs[0]
+		switch {
+		case config.ResourceManager.AgentRM != nil:
+			return agentrm.New(db, echo, config, opts, cert)
+		case config.ResourceManager.KubernetesRM != nil:
+			return kubernetesrm.New(db, config, tcd, opts, cert)
+		default:
+			panic("no expected resource manager config is defined")
+		}
+	}
+
+	// Set the default RM name for the multi-rm, if the default indexed config is already named.
+	// This guarantees that any multi-rm request is sent to the default RM if a given RM name
+	// doesn't exist.
+	defaultRMName := config.DefaultRMName
+	if name := rmConfigs[config.DefaultRMIndex].ResourceManager.Name(); name != "" {
+		defaultRMName = name
+	}
+
+	rms := map[string]rm.ResourceManager{}
+
+	for idx, cfg := range rmConfigs {
+		c := cfg
+		switch {
+		case c.ResourceManager.AgentRM != nil:
+			// If the default config index doesn't have a named RM, name it.
+			if idx == config.DefaultRMIndex && c.ResourceManager.Name() == "" {
+				c.ResourceManager.AgentRM.Name = defaultRMName
+			}
```

Comment:
shouldn't this be able to be defaulted by the config pkg? and if it isn't by here we just return an error?

Reply:
@NicholasBlaskey ? my understanding is the opposite
also, I updated the code to not rewrite the RM name in the config, but rather to just substitute the defaultRMName in to the map if it's blank in the default config index

Reply:
We should validate the config is set for every resource manager
https://github.com/determined-ai/determined/blob/63adae5e933cde753b95d2592357b82e39ca370f/master/internal/config/resource_manager_config.go#L224

If a user doesn't provide a name for the default resource manager, we will default it for them

I guess my comment here is that the config package should have responbility for making the config set, other packages shouldn't assume that the config is invalid or try to apply defaults to it.